src/threads.rs

```diff
@@ -0,0 +1,303 @@
+            threads: threads,
+            terminated_threads: Default::default(),
+        }
+    }
+}
+
+impl<'mir, 'tcx: 'mir> ThreadSet<'mir, 'tcx> {
+    /// Borrow the stack of the active thread.
+    fn active_thread_stack(&self) -> &[Frame<'mir, 'tcx, Tag, FrameData<'tcx>>] {
+        &self.threads[self.active_thread].stack
+    }
+    /// Mutably borrow the stack of the active thread.
+    fn active_thread_stack_mut(&mut self) -> &mut Vec<Frame<'mir, 'tcx, Tag, FrameData<'tcx>>> {
+        &mut self.threads[self.active_thread].stack
+    }
+    /// Create a new thread and returns its id.
+    fn create_thread(&mut self) -> ThreadId {
+        let new_thread_id = ThreadId::new(self.threads.len());
+        self.threads.push(Default::default());
+        new_thread_id
+    }
+    /// Set an active thread and return the id of the thread that was active before.
+    fn set_active_thread(&mut self, id: ThreadId) -> ThreadId {
+        let active_thread_id = self.active_thread;
+        self.active_thread = id;
+        assert!(self.active_thread.index() < self.threads.len());
+        active_thread_id
+    }
+    /// Get the id of the currently active thread.
+    fn get_active_thread(&self) -> ThreadId {
+        self.active_thread
+    }
+    /// Mark the thread as detached, which means that no other thread will try
+    /// to join it and the thread is responsible for cleaning up.
+    fn detach_thread(&mut self, id: ThreadId) {
+        self.threads[id].detached = true;
+    }
+    /// Mark that the active thread tries to join the thread with `joined_thread_id`.
+    fn join_thread(&mut self, joined_thread_id: ThreadId) {
+        assert!(!self.threads[joined_thread_id].detached, "Bug: trying to join a detached thread.");
+        assert_ne!(joined_thread_id, self.active_thread, "Bug: trying to join itself");
+        assert!(
+            self.threads
+                .iter()
+                .all(|thread| thread.state != ThreadState::Blocked(joined_thread_id)),
+            "Bug: multiple threads try to join the same thread."
+        );
+        if self.threads[joined_thread_id].state != ThreadState::Terminated {
+            // The joined thread is still running, we need to wait for it.
+            self.threads[self.active_thread].state = ThreadState::Blocked(joined_thread_id);
+            trace!(
+                "{:?} blocked on {:?} when trying to join",
+                self.active_thread,
+                joined_thread_id
+            );
+        }
+    }
+    /// Get ids of all threads ever allocated.
+    fn get_all_thread_ids(&mut self) -> Vec<ThreadId> {
+        (0..self.threads.len()).map(ThreadId::new).collect()
+    }
+    /// Decide which thread to run next.
+    ///
+    /// Returns `false` if all threads terminated.
+    fn schedule(&mut self) -> InterpResult<'tcx, bool> {
+        if self.threads[self.active_thread].check_terminated() {
+            // Check if we need to unblock any threads.
```

Comment:
http://man7.org/linux/man-pages/man3/pthread_exit.3.html

> When a thread terminates, process-shared resources (e.g., mutexes,
>        condition variables, semaphores, and file descriptors) are not
>        released, and functions registered using atexit(3) are not called.

Reply:
@bjorn3 I do not understand your comment. What exactly you hand in mind?